Rich Text Editor Blocks v16 Documentation (#7305)
umbraco/UmbracoDocs: 16/umbraco-cms/fundamentals/backoffice/property-editors/built-in-umbraco-property-editors/rich-text-editor/blocks.md

Blocks can be added to the Rich Text Editor when:

*   Available Blocks are specified as part of the Rich Text Editor Data Type configuration
*   The **Insert Block** toolbar option is enabled in the Rich Text Editor

![RTE Insert Block Toolbar Button](images/rte-blocks-toolbar-insert-button.png)

## Configure Blocks

Blocks are Element Types that must be created before configuring them as blocks in the Rich Text Editor. You can create Element Types in the **Settings** > **Document Types** section

Blocks functionality can then be configured through the Rich Text Editor Data Type configuration as follows.

1. Navigate to **Settings** > **Data Types**.
2. Select your Rich Text Editor Data Type or create a new one.
3. In the **Available Blocks** section, add the Element Types you want to make available as blocks.
4. Configure each Block Type with the options described below.

![RTE Blocks Data Type Configuration](images/rte-blocks-datatype-configuration.png)

### Editor Appearance

Configure how blocks appear and behave in the Content section:

> *   **Label** - Define how the block appears in the editor. Umbraco 16 uses UFM (Umbraco Flavored Markdown) syntax for dynamic labels. Use `{=propertyAlias}` to display property values (e.g., `{=author}` for a text property containing the name of an author, or `Written by: {=author}` for a label with static text and dynamic content)

Reviewer: Link to the Umbraco Flavored Markdown page :)

# Blocks

Blocks enable editors to insert structured content elements directly into the Rich Text Editor (RTE). Blocks are [Element Types](../../../../../../data/defining-content/#element-types) and can be configured with custom properties, styling, and behavior.

Blocks can be added to the Rich Text Editor when:

> *   Available Blocks are specified as part of the Rich Text Editor Data Type configuration

Reviewer 2: Extra spaces between bullet an text is not necessary 🙏 

Could you please go through the PR an remove these extra spaces?

```suggestion
* Available Blocks are specified as part of the Rich Text Editor Data Type configuration
```

*   The **Insert Block** toolbar option is enabled in the Rich Text Editor

![RTE Insert Block Toolbar Button](images/rte-blocks-toolbar-insert-button.png)

## Configure Blocks

Blocks are Element Types that must be created before configuring them as blocks in the Rich Text Editor. You can create Element Types in the **Settings** > **Document Types** section

Blocks functionality can then be configured through the Rich Text Editor Data Type configuration as follows.

1. Navigate to **Settings** > **Data Types**.
2. Select your Rich Text Editor Data Type or create a new one.

> 3. In the **Available Blocks** section, add the Element Types you want to make available as blocks.
> 4. Configure each Block Type with the options described below.

Reviewer 2: Always start each step in an ordered list with the "action" word.

```suggestion
3. Locate the **Available Blocks** section.
4. Add the Element Types you want to make available as blocks.
5. Configure each Block Type with the options described below.
```

![RTE Blocks Editor Appearance Settings](images/rte-blocks-editor-appearance-settings.png)

### Data Models

Configure the content structure for your blocks:

*   **Content model** - The Element Type that defines the main content properties for the block (required)
*   **Settings model** - Optional Element Type that defines additional settings or configuration options for the block

![RTE Blocks Data Models Settings](images/rte-blocks-data-models-settings.png)

### Catalogue Appearance

Control how blocks appear in the block picker:

*   **Background color** - Background color displayed behind the block icon or thumbnail
*   **Icon Color** - Color of the Element Type icon
*   **Thumbnail** - Custom image to replace the default Element Type icon

## Working with Blocks

### Adding Blocks to Content

Editors can add blocks to rich text content using the **Insert Block** toolbar button:

1. Position the cursor where you want to insert the block
2. Click the **Insert Block** button in the Rich Text Editor toolbar
3. Select the desired Block from the available options
4. Configure the block content (and settings, if provided)

> 5. The block appears in the editor as a structured element

Reviewer 2: ```suggestion

The block appears in the editor as a structured element
```

![Adding Blocks to Content - Step 1](images/rte-blocks-adding-to-content-1.png)

![Adding Blocks to Content - Step 2](images/rte-blocks-adding-to-content-2.png)

> ![Adding Blocks to Content - Step 3](images/rte-blocks-adding-to-content-3.png)

Reviewer 2: Could you add these images to the list instead?
That way you see what you need to do, when you need to do it.

Add them to the list by following this format:

1. Instruction.

![image text](image.png)

2. More instructions.

Hope it makes sense 🙏

        </a>
    </div>
}
```

### Type-Safe Rendering with Models Builder

When using Models Builder, specify the Content and Settings models for type-safe access:

```csharp
@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage<Umbraco.Cms.Core.Models.Blocks.RichTextBlockItem<QuoteBlock, QuoteBlockSettings>>
@using ContentModels = Umbraco.Cms.Web.Common.PublishedModels;

@{
    var style = "";
    if (Model.Settings?.Color != null)
    {
        style = $"style=\"border-left-color: {Model.Settings.Color};\"";
    }
}

<blockquote class="quote-block" @Html.Raw(style)>
    @if (!string.IsNullOrEmpty(Model.Content.QuoteText))
    {
        <p class="quote-text">@Html.Raw(Model.Content.QuoteText)</p>
    }

    @if (!string.IsNullOrEmpty(Model.Content.Citation))
    {
        <cite class="quote-citation">@Model.Content.Citation</cite>
    }
</blockquote>
```

## Build a Custom Backoffice View

Building Custom Views for Block representations in Backoffice is the same for all Block Editors. [Read about building a Custom View for Blocks here](../../../../../tutorials/creating-custom-views-for-blocklist.md).

## Best Practices

### Content Design

> *   Design blocks for reusable content patterns

Reviewer 2: Let's keep it consistent and end each list item with a dot.

```suggestion
* Design blocks for reusable content patterns.
```

*   Keep block content focused on a single purpose
*   Use descriptive labels that help editors understand the block's function

### Performance

*   Avoid creating too many Blocks - this can overwhelm content editors
*   Use appropriate caching strategies for block rendering
*   Consider the impact of complex blocks on editor performance

### Accessibility

*   Ensure block markup follows accessibility guidelines
*   Provide meaningful labels and descriptions
*   Test block rendering with screen readers

## Related Articles

> *   [Element Types](../../../../../../data/defining-content/#element-types)

Reviewer 2: Broken link 🤔 This is an attempt to fix the link.

```suggestion
*   [Element Types](../../../../data/defining-content/default-document-types.md#element-types)
```

> *   [Rich Text Editor Configuration](../configuration.md)
> *   [Rich Text Editor Extensions](../extensions.md)

Reviewer 2: Their within the same folder at the this article, so no path is needed 🙏 

```suggestion
*   [Rich Text Editor Configuration](configuration.md)
*   [Rich Text Editor Extensions](extensions.md)
```
